File: src/main/java/com/rtomyj/yugiohAPI/controller/BanListDatesController.java
package com.rtomyj.yugiohAPI.controller;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.servlet.http.HttpServletRequest;

import com.rtomyj.yugiohAPI.helper.LogHelper;
import com.rtomyj.yugiohAPI.model.BanLists;
import com.rtomyj.yugiohAPI.service.BanListService;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import io.swagger.annotations.Api;
import io.swagger.annotations.ApiOperation;
import io.swagger.annotations.ApiResponse;
import io.swagger.annotations.ApiResponses;

@RequestMapping(path = "${ygo.endpoints.ban-list-dates-v1}", produces = "application/json; charset=utf-8")
@RestController
@CrossOrigin(origins = "*")
@Api(description = "Request information about current and past ban lists", tags = "Ban List")
public class BanListDatesController
{
	@Autowired
	private BanListService banListService;

	@Autowired
	@Value("${ygo.endpoints.ban-list-dates-v1}")
	private String endPoint;

	@Autowired
	private HttpServletRequest httpRequest;

	private static final Logger LOG = LogManager.getLogger();

	private static final Map<String, List<BanLists>> BAN_LISTS_START_DATES_CACHE = new HashMap<>();



	/**
	 * @return item
	 */
	@GetMapping()
	@ApiOperation(value = "Get dates of ban lists stored in database", response = ResponseEntity.class, tags = "Ban List")
	@ApiResponses( value = {
		@ApiResponse(code = 200, message = "OK")
	}
	)
	public ResponseEntity<Map<String, List<BanLists>>> startDatesOfBanLists()
	{
		if (BAN_LISTS_START_DATES_CACHE.size() == 0)
		{
			List<BanLists> banStartDates = (ArrayList<BanLists>) banListService.getBanListStartDates();
			BAN_LISTS_START_DATES_CACHE.put("banListStartDates", banStartDates);
		}


		HttpStatus status = HttpStatus.OK;
		LOG.info(LogHelper.requestInfo(httpRequest.getRemoteHost(), endPoint, String.format("Responding with { %s }", status)));
		return new ResponseEntity<>(BAN_LISTS_START_DATES_CACHE, status);
	}
}

File: src/main/kotlin/com/rtomyj/skc/find/BanListDatesController.kt
package com.rtomyj.skc.find

import com.rtomyj.skc.config.ReactiveMDC
import com.rtomyj.skc.model.BanListDates
import com.rtomyj.skc.util.constant.SwaggerConstants
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.tags.Tag
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Mono

/**
 * Configures endpoint(s) for returning user the dates of the ban lists in the database.
 */
@RestController
@RequestMapping(path = ["/ban_list/dates"], produces = ["application/json; charset=UTF-8"])
@Tag(name = SwaggerConstants.BAN_LIST_TAG_NAME)
class BanListDatesController
/**
 * Create object instance.
 * @param banListDatesService Service object to use to accomplish functionality needed by this endpoint.
 */ @Autowired constructor(
  /**
   * Service object used to interface the database DAO
   */
  val banListDatesService: BanListDatesService
) {

  companion object {
    @JvmStatic
    private val log = LoggerFactory.getLogger(this::class.java.name)
  }


  /**
   * Looks in the database for the start dates of all ban lists stored in database.
   * @return Map that contains a list of all dates of the ban lists in database.
   */
  @ApiResponse(responseCode = "200", description = SwaggerConstants.HTTP_200_SWAGGER_MESSAGE)
  @ApiResponse(responseCode = "422", ref = "Unprocessable Entity")
  @ApiResponse(responseCode = "500", ref = "Internal Server Error")
  @Operation(
    summary = "Retrieve start (effective) dates of all ban lists stored in database in logical order. These dates are \"valid\" start dates that can be used by other endpoints ban list endpoints.",
    tags = [SwaggerConstants.BAN_LIST_TAG_NAME]
  )
  @GetMapping
  fun banListStartDates(
    @RequestParam(
      name = "format", required = true, defaultValue = "TCG"
    ) format: String = "TCG"
  ): Mono<BanListDates> = ReactiveMDC.deferMDC(banListDatesService
      .retrieveBanListStartDates(format)
      .doOnSuccess {
        log.info(
          "Successfully retrieved all effective start dates for ban list using format {}. Currently there are {} ban lists",
          format,
          it.dates.size
        )
      }
      .doOnSubscribe {
        log.info("Retrieving all effective start dates for ban lists using format {}.", format)
      })
}